proposed/2022/PrivateAssetIndepdentFromExcludeAsset.md

```diff
@@ -0,0 +1,231 @@
+    <dependencies>
+      <group targetFramework=".NETStandard2.0">
+        <dependency id="Microsoft.Windows.CsWin32" version="0.2.138-beta" exclude="Runtime,Compile,Native,Analyzers,BuildTransitive" />
+      </group>
+    </dependencies>
+```
+
+After change nuspec file:
+
+```.net
+    <dependencies>
+      <group targetFramework=".NETStandard2.0">
+        <dependency id="Microsoft.Windows.CsWin32" version="0.2.138-beta" include="All" />
+      </group>
+    </dependencies>
+```
+
+##### Case 2 for PackageReference
+
+Package reference in csproj file.
+
+```.net
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <VersionSuffix>beta</VersionSuffix>
+    <PrivateAssetIndependent>True</PrivateAssetIndependent>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" PrivateAssets="none" IncludeAssets="none" />    
+  </ItemGroup>
+```
+
+Before change nuspec file:
+
+```.net
+    <dependencies>
+      <group targetFramework=".NETStandard2.0" />
+    </dependencies>
+```
+
+After change nuspec file:
+
+```.net
+    <dependencies>
+      <group targetFramework=".NETStandard2.0">
+        <dependency id="Microsoft.Windows.CsWin32" version="0.2.138-beta" include="All" />
+      </group>
+    </dependencies>
+```
+
+##### Case for Project to Project reference
+
+`ProjectReference` in parent `ClassLibrary1` csproj file:
+
+```.net
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\refissue\refissue.csproj" />
+  </ItemGroup>
+```
+
+And `PackageReference` in `refissue.csproj` file:
+
+```.net
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <VersionSuffix>beta</VersionSuffix>
+    <PrivateAssetIndependent>True</PrivateAssetIndependent>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" PrivateAssets="none" ExcludeAssets="all" />
+  </ItemGroup>
+```
+
+Before changing, `Microsoft.Windows.CsWin32.props` file in `build` asset doesn't flow to ClassLibrary1.csproj.nuget.g.props file in obj folder.
+
+After change, `Microsoft.Windows.CsWin32.props` file in `build` asset flow to ClassLibrary1.csproj.nuget.g.props file in obj folder.
+
+### Technical explanation
+
+We already have a [logic](hhttps://github.com/NuGet/NuGet.Client/blob/380415d812681ebf1c8aa0bc21533d4710514fc3/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs#L577-L582) that combines `IncludeAssets/ExcludeAssets` options with `PrivateAssets` for pack operation, we just make logic depending on opt-in property.
+
+## Drawbacks
+
+<!-- Why should we not do this? -->
+- Might break some consuming projects, that is why it's opt-in feature. See table in #functional-explanation above for more details.
+
+## Rationale and alternatives
```

Comment:
How about an alternative of adding a new metadata on PackageReference items?

Something like `TransitiveAssets` or `ExcludeTransitiveAssets`?  I think this makes it much easier for a customer who has never read the docs or github issue on the feature could guess what it does.

Example 1 would then change to `<PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" IncludeAssets="build" ExcludeTransitiveAssets="none" />`

Reply:
Sure, added it. Previously discussed about it [here](https://github.com/NuGet/Home/pull/12313#discussion_r1043079630).

Reply:
The comment you linked to is about something different. It's asking if the opt-in should be via item metadata or a project property, but it's otherwise maintaining the same fundamental design of changing how Include/Private/Exclude Assets are used to calculate package dependency asset flow via an opt-in flag.

My proposal avoids needing opt-in, or changing existing behaviour. Instead, a new keyword will set the asset transitive flow when specified.

I'm trying to think what's easier for customers who are not multi-year experts of .NET and MSBuild to understand. Basically, "What would be easier to document" is the same question. Also, what will be easier for us to support when customers create bug reports. We very often ask customers for more information, or for a sample reproduction, when customers report problems. If we see a bug report with `<PackageReference TransitiveAssets="..."` we immidiately know the customer is using this feature. Many customers don't know that `Directory.Build.props` exist, so wouldn't know to check before filing a bug report, or including it a bug report.

There is at least one (significant?) drawback to the idea though. @AArnott said he wanted to opt in entire repos, which is easy with the project property opt-in. But if transitive flow is specified by a new item metadata, then every PackageReference that wants to change the default will need to be updated. However, this is just one customer feedback. There was one other customer who engaged in the home issue, but their issue was resolved using existing features, so they don't need this new feature after all. So, I don't know if customers will want to opt-in entire projects or specific PackageReferences. A sample size of 1 isn't enough to generalize to all .NET developers. Having said that, the feature request only got 6 upvotes, so it's probably a feature that isn't going to be widely used. Maybe a sample size of one is representative of the other 5.

Reply:
Hopefully I captured your concern correctly [here](https://github.com/NuGet/Home/pull/12313/commits/72bcb2c814896d1c8ab076087762a8865a35fa2f). I'll keep this discussion open until others return from holidays.

Reply:
@NuGet/nuget-client 
Just case if folks are interested, I can show difference between 2 approaches:

IncludeAssets|PrivateAssets|Flows transitively (default)|Flows transitively (PrivateAssetIndependent=true)|Flows transitively (TransitiveAssets="build")
--|--|--|--|--
☑️ | 🔲 | ✅ | ✅ | ✅
☑️ | ☑️ | ❌ | ❌ | ✅
🔲 | 🔲 | ❌ | ✅ | ✅
🔲 | ☑️ | ❌ | ❌ | ✅

1. Currently newly added `PrivateAssetsIndependent` property, we only set `<PrivateAssetIndependent>True</PrivateAssetIndependent>` once in project/solution to opt-into the experience, we don't change existing `PackageReference`. Please see `column 4` in the above table.

    Upside 
    - With this approach we don't need much work for opt-in the existing solution and get consistent expectation for asset flow.
    - If make this one default experience without opt-in later then it's easy.
    
    Downside
    - No `packagereference` level fine control but considering @AArnott wanted to opt-in entire repos it's fine.
    - Customer still need to know how IncludeAsset and ExcudeAsset combination works. Personally, I feel it's not big issue, it's simple boolean [and](https://github.com/NuGet/NuGet.Client/blob/5b0ee2562df6cf2c66cc9905a8a65f2b09ea5473/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs#L129) operation between them, probably we just need to improve our docs.

2. Alternative would be adding yet another metadata `TransitiveAssets` which like existing `IncludeAssets/ExcludeAssets` metadata, only for control transitive asset flow. Please see `column 5` in the above table.

    Upside 
    - We get `PackageReference` level fine control.
    - Explicitly set which type of assets can flow. `<PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" IncludeAssets="build" TransitiveAssets="compile;build,analyzers" />`
    
    Downside
    - Hard to opt-in existing repo/solution, need to add this all the packagereferences manually. But it's easy for opt-in new ones.
    - We get consistent expectation for `packagereferences` opted in with this new metadata. But for non-opted-in case in below case for `transitive` assets still flows to parent project for packagereferences (IncludeAsset=☑️, PrivateAssets=🔲). So that means we can do same thing 2 different ways, we may not have consistent control over which asset flows, unless we may have to pay attention on how IncludeAsset and ExcudeAsset combination works, technically it's same downside we mentioned in above approach. 

![image](https://user-images.githubusercontent.com/8766776/210456271-5453d900-44c9-41b1-bfc9-46946b7b6e5d.png)